File: app/src/main/java/cz/johnyapps/cheers/adapters/BaseAdapter.java
package cz.johnyapps.cheers.adapters;

import android.content.Context;
import android.view.LayoutInflater;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.RecyclerView;

public abstract class BaseAdapter<VIEW_HOLDER extends RecyclerView.ViewHolder, DATA> extends RecyclerView.Adapter<VIEW_HOLDER> {
    @NonNull
    private final Context context;
    @NonNull
    private final LayoutInflater inflater;

    public BaseAdapter(@NonNull Context context) {
        this.context = context;
        this.inflater = LayoutInflater.from(context);
    }

    public void update(@Nullable DATA data) {
        onUpdate(data);
        notifyDataSetChanged();
    }

    protected abstract void onUpdate(@Nullable DATA data);

    @NonNull
    public Context getContext() {
        return context;
    }

    @NonNull
    public LayoutInflater getInflater() {
        return inflater;
    }
}

File: app/src/main/java/cz/johnyapps/cheers/adapters/BaseAdapter.kt
package cz.johnyapps.cheers.adapters

import android.content.Context
import android.view.View
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView

abstract class BaseAdapter<T, VH: BaseAdapter<T, VH>.ViewHolder>(itemCallback: DiffUtil.ItemCallback<T>): ListAdapter<T, VH>(itemCallback) {
    var onListChangeListener: OnListChangedListener? = null

    fun getContext(viewHolder: VH): Context {
        return viewHolder.itemView.context
    }

    override fun onCurrentListChanged(previousList: MutableList<T>, currentList: MutableList<T>) {
        super.onCurrentListChanged(previousList, currentList)
        onListChangeListener?.onChange(currentList.isEmpty())
    }

    interface OnRootClickListener {
        fun onClick(view: View)
    }

    interface OnListChangedListener {
        fun onChange(isListEmpty: Boolean)
    }

    open inner class ViewHolder(itemView: View): RecyclerView.ViewHolder(itemView) {
        private val rootOnClickListeners = ArrayList<OnRootClickListener>()

        init {
            itemView.setOnClickListener {
                for (listener in rootOnClickListeners) {
                    listener.onClick(it)
                }
            }
        }

        fun addRootOnClickListener(listener: OnRootClickListener) {
            rootOnClickListeners.add(listener)
        }
    }
}